### PR TITLE
part 1 of plan eval. 

### PR DESCRIPTION
**Pull request by siyuanfeng-tri**
0. Added a wrapper for PiecewisePolynomial and up to its second derivative. 
1. Added a wrapper for Cartesian trajectory, the position part is a PP, and the orientation is a PiecewiseSlerp.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/5585)

### CODE REVIEW

**Comment by siyuanfeng-tri** [comment]
+@naveenoid for feature review. should be pretty straight forward. thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-KfnXCZGBh0TtBMz_WmO:b-b1s00d)*

**Comment by naveenoid** [comment]
:lgtm:

---

Reviewed 14 of 16 files at r1.
Review status: 14 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfv8uB4P_TNRgWxNTak:-Kfv8uB4P_TNRgWxNTal:b5m5bbu) ([raw file](https://github.com/robotlocomotion/drake/blob/4bc25e6edc423b3d1ee6165ddabdf4e32f97fb11/drake/common/trajectories/piecewise_quaternion.cc#L23)):*
> ```cpp
>     // A quick reference:
>     // https://fgiesen.wordpress.com/2013/01/07/
>     // small-note-on-quaternion-distance-metrics/
> ```

BTW:  Perhaps eventually we better move this cited documentation to drake spinx? Im not totally comfortable with a wordpress ref. The underlying distance measure is simple in anycase.

---

*[drake/examples/QPInverseDynamicsForHumanoids/param_parsers/param_parser.h, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-KfvBXLYYd_n5m5OAqML:-KfvBXLYYd_n5m5OAqMM:b8iheh2) ([raw file](https://github.com/robotlocomotion/drake/blob/4bc25e6edc423b3d1ee6165ddabdf4e32f97fb11/drake/examples/QPInverseDynamicsForHumanoids/param_parsers/param_parser.h#L313)):*
> ```objc
>   /**
>    * Returns a QpInput for the given contacts and tracked bodies using the
>    * parameters hold by this instance. Note that this function only sets the
> ```

BTW: nit: *held

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-KfvFAua-ZLJU7-bdhFp:bnfp4nl)*

**Comment by siyuanfeng-tri** [comment]
Review status: 12 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.cc, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfv8uB4P_TNRgWxNTak:-KfvV_6QSR8VIKo64Mul:b-bfvhvq) ([raw file](https://github.com/robotlocomotion/drake/blob/4bc25e6edc423b3d1ee6165ddabdf4e32f97fb11/drake/common/trajectories/piecewise_quaternion.cc#L23)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW:  Perhaps eventually we better move this cited documentation to drake spinx? Im not totally comfortable with a wordpress ref. The underlying distance measure is simple in anycase.
</blockquote></details>

o well, i changed it to matt mason's lecture notes. 

---

*[drake/examples/QPInverseDynamicsForHumanoids/param_parsers/param_parser.h, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-KfvBXLYYd_n5m5OAqML:-KfvIA_WJuw1fyepdxUO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/4bc25e6edc423b3d1ee6165ddabdf4e32f97fb11/drake/examples/QPInverseDynamicsForHumanoids/param_parsers/param_parser.h#L313)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: nit: *held
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by siyuanfeng-tri** [comment]
+@jwnimmer-tri for platform review. thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-KfvVdzGKRMz6Xe7GRl_:b-uvnfud)*

**Comment by jwnimmer-tri** [comment]
Please double-check that all changes have test coverage.  Some `foo.cc` files appear to have changed without any changes to `test/foo_test.cc`.  Perhaps the tests come under different filenames (and I'll be doing a more detailed pass to confirm), but I thought I'd post early before I dig in, in case there are gaps that you want to start in on immediately.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-KfwbykiTWszl32mx972:bn0k9s7)*

**Comment by siyuanfeng-tri** [comment]
ur, most of it. i did intentionally skipped testing stuff for like makeAffine( ), and PiecewiseFunction::isTimeInRange(double time) because they are trivial changes. 
i will add a test for the new MakeQpInput(). 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-Kfwe75UGbxrzHPqIDIm:b-9i648l)*

**Comment by jwnimmer-tri** [comment]
Checkpoint.

---

Reviewed 11 of 16 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfwo_xtnljFKF8LomK9:-Kfwo_xuB-LnVA81kIdd:b-p3wpp8) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/common/trajectories/piecewise_quaternion.h#L106)):*
> ```objc
>    * quaternion knot points are within @p tol.
>    */
>   bool is_approx(const PiecewiseQuaternionSlerp<Scalar>& other, Scalar tol)
> ```

That `tol` is a `Scalar` implies that it could be `AutoDiff` something.  `AutoDiff` values must never be passed by-value because of alignment issues.

---

*[drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc, line 141 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfwv2u-B2URarbMjOOn:-Kfwv2u-B2URarbMjOOo:b4jq0z5) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc#L141)):*
> ```cpp
>       H1.linear() = kFootToSensorRotationOffset.transpose();
>       H1.translation() = -kFootToSensorPositionOffset + kFootToSoleOffset;
>       H1.makeAffine();
> ```

_Meta_ If these (three) additions of `makeAffine` were important enough to fix, then the default presumption is that they are also important enough to test.  Is there any test coverage?  (Or convince me that there's some reason not to test them, e.g., they are only for computational performance improvements and not correctness.)

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/BUILD, line 5 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-KfwuJskMM1aWUWi6Z4H:-KfwuJslZcDrPrj_8jeg:b-407oya) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/BUILD#L5)):*
> ```Python
> 
> load("//tools:cpplint.bzl", "cpplint")
> load("//tools:drake.bzl", "drake_cc_binary", "drake_cc_googletest", "drake_cc_library")
> ```

BTW Please wrap to 80 cols (see other similar BUILD files for examples of how to line wrap this).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-KfwviJlWDW5kBVU1R2p:by96gdd)*

**Comment by siyuanfeng-tri** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc, line 141 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfwv2u-B2URarbMjOOn:-KfwwiMyVvs0JredQv9P:bxkfpx3) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc#L141)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ If these (three) additions of `makeAffine` were important enough to fix, then the default presumption is that they are also important enough to test.  Is there any test coverage?  (Or convince me that there's some reason not to test them, e.g., they are only for computational performance improvements and not correctness.)
</blockquote></details>

so this is a very disturbing but relatively new discovery i made yesterday although i have being using Isometry3d for years.... 
when you make a Isometry and do the normal transformation operations operator * / inverse, the last row doesn't matter because of the implementation. if those were done with the normal 4 by 4 matrix operation, i am really screwed. 

this only came up the other day when i was doing element wise comparison between two Isometry3d, one was properly initialized with the last row being 0 0 0 1, and the other doesn't have the last row initialized correctly. 

i do want to make sure everything Isometry3d in drake is initialized properly, but i don't know how to do it properly. because you can always do what did here, and never realized this. and i don't know where to put the tests. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by jwnimmer-tri** [comment]
*[drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc, line 141 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfwv2u-B2URarbMjOOn:-Kfxd7rAWv4qHVqso19m:b2wlqs) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc#L141)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

so this is a very disturbing but relatively new discovery i made yesterday although i have being using Isometry3d for years.... 
when you make a Isometry and do the normal transformation operations operator * / inverse, the last row doesn't matter because of the implementation. if those were done with the normal 4 by 4 matrix operation, i am really screwed. 

this only came up the other day when i was doing element wise comparison between two Isometry3d, one was properly initialized with the last row being 0 0 0 1, and the other doesn't have the last row initialized correctly. 

i do want to make sure everything Isometry3d in drake is initialized properly, but i don't know how to do it properly. because you can always do what did here, and never realized this. and i don't know where to put the tests. 
</blockquote></details>

Well, its true that `Update` is a big method that does lots of things at once, so is difficult to unit test.

What if we had something like this?
```c++
// math/something.h
template <typename T>
drake::MakeIsometry3(
    const Matrix3<T>& linear,
    const Vector3<T>& translation) {
  Isometry3<T> result;
  result.linear() = linear;
  result.translation() = translation;
  result.makeAffine();
  return result;
  // ... or maybe this method should be calling
  //   Isometry<T>::fromPositionOrientationScale
  // instead?
}

// Update
 auto H1 = MakeIsometry3(
    kFootToSensorRotationOffset.transpose(),
    -kFootToSensorPositionOffset + kFootToSoleOffset);
```

In other words, if making an `Isometry3d` is really such a landmine, we should wrap the fix in a well-tested function and then call that function everywhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by siyuanfeng-tri** [comment]
Review status: 10 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/trajectories/piecewise_quaternion.h, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfwo_xtnljFKF8LomK9:-Kg-TDa402NFQ-N9AQQM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/common/trajectories/piecewise_quaternion.h#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

That `tol` is a `Scalar` implies that it could be `AutoDiff` something.  `AutoDiff` values must never be passed by-value because of alignment issues.
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc, line 141 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kfwv2u-B2URarbMjOOn:-Kg-YJtFH3-0OH5bsVes:b-sl3x58) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/examples/QPInverseDynamicsForHumanoids/humanoid_status.cc#L141)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Well, its true that `Update` is a big method that does lots of things at once, so is difficult to unit test.

What if we had something like this?
```c++
// math/something.h
template <typename T>
drake::MakeIsometry3(
    const Matrix3<T>& linear,
    const Vector3<T>& translation) {
  Isometry3<T> result;
  result.linear() = linear;
  result.translation() = translation;
  result.makeAffine();
  return result;
  // ... or maybe this method should be calling
  //   Isometry<T>::fromPositionOrientationScale
  // instead?
}

// Update
 auto H1 = MakeIsometry3(
    kFootToSensorRotationOffset.transpose(),
    -kFootToSensorPositionOffset + kFootToSoleOffset);
```

In other words, if making an `Isometry3d` is really such a landmine, we should wrap the fix in a well-tested function and then call that function everywhere.

</blockquote></details>

hmm switched to fromPositionOrientationScale. 

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/BUILD, line 5 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-KfwuJskMM1aWUWi6Z4H:-Kg-Tgwu1d_RrWwsWZr_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bb8dd928474fd80fc5d25ee0941f94f455aa872e/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/BUILD#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Please wrap to 80 cols (see other similar BUILD files for examples of how to line wrap this).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by jwnimmer-tri** [comment]
Reviewed 1 of 16 files at r1, 4 of 8 files at r3.
Review status: 14 of 18 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by jwnimmer-tri** [comment]
Reviewed 2 of 8 files at r3.
Review status: 16 of 18 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg0-6u8CARPdLsQiaK1:-Kg0-6u8CARPdLsQiaK2:b-vv7twh) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L1)):*
> ```objc
> #pragma once
> ```

_Meta_ `kcov` shows that this file has only like 50% of the methods ever called from any automated tests.  Please add tests.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-quGNKsNNDxRQo3f1:-Kg-quGOyQhTbhRVMsMK:b-mfuf8t) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L20)):*
> ```objc
>  * here.
>  */
> typedef std::set<const RigidBody<double>*> ContactState;
> ```

FYI `set` usually means that you care about the iteration order, vs `unordered_set` which means that you only care about membership.  For pointers, the latter is most always a better match.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg0-_SpQT3w1xfcTy2Y:-Kg0-_SpQT3w1xfcTy2Z:bpmh3bp) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L31)):*
> ```objc
>  */
> template <typename T>
> class PiecewiseCubicTrajectory {
> ```

_Meta_ in this whole file (and maybe other files in this PR), we need to mark `DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN` or similar on every class.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-vYHj6oqsvti8J0rM:-Kg-vYHj6oqsvti8J0rN:b-6x9rqw) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L44)):*
> ```objc
>     q_ = position_traj;
>     qd_ = q_.derivative();
>     qdd_ = qd_.derivative();
> ```

We shouldn't be `operator=` assigning `PiecewisePolynomial` because is a non-final virtual class!  Can you add the `final` modifier to that class in this PR?  Or at least add a TODO that the polynomial stuff is fundamentally broken in its use of C++ and should be repaved?

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 107 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-xrxuEP2adpW1B2NF:-Kg-xrxuEP2adpW1B2NG:bkn5o4z) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L107)):*
> ```objc
>    */
>   const PiecewisePolynomial<T>& get_acceleration_trajectory() const {
>     return qd_;
> ```

This should be `qdd_`?
If so, then a unit test is missing.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 110 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-zQAe17tfTQj4hwGL:-Kg-zQAe17tfTQj4hwGM:b-fkmb6r) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L110)):*
> ```objc
>   }
> 
>   bool operator==(const PiecewiseCubicTrajectory<T>& other) const {
> ```

_Meta_ I don't think overloading `operator==` for anything other than exact equality is allowed by https://google.github.io/styleguide/cppguide.html#Operator_Overloading.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by siyuanfeng-tri** [comment]
Review status: 15 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg0-6u8CARPdLsQiaK1:-Kg0mhkxNmYokPB85515:ba7t2sc) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ `kcov` shows that this file has only like 50% of the methods ever called from any automated tests.  Please add tests.
</blockquote></details>

you mean plan_eval_utils.h? it seems fine to me, maybe i read it wrong? the other stuff seems pretty bad.. 

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-quGNKsNNDxRQo3f1:-Kg0L0iKQ5zBSAw8_3Xm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L20)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `set` usually means that you care about the iteration order, vs `unordered_set` which means that you only care about membership.  For pointers, the latter is most always a better match.
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg0-_SpQT3w1xfcTy2Y:-Kg0LP255W5qiRPc0ZkC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L31)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ in this whole file (and maybe other files in this PR), we need to mark `DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN` or similar on every class.
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-vYHj6oqsvti8J0rM:-Kg0R96az-2h9KILGNaN:banoi2a) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

We shouldn't be `operator=` assigning `PiecewisePolynomial` because is a non-final virtual class!  Can you add the `final` modifier to that class in this PR?  Or at least add a TODO that the polynomial stuff is fundamentally broken in its use of C++ and should be repaved?
</blockquote></details>

added final and default copyable stuff to PiecewiseQuaternionSlerp and PiecewisePolynomial.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 107 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-xrxuEP2adpW1B2NF:-Kg0NowVxaJ8dGFX-LRy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L107)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This should be `qdd_`?
If so, then a unit test is missing.
</blockquote></details>

Done.

---

*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 110 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg-zQAe17tfTQj4hwGL:-Kg0QqqxmjRiaVlKl5hf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ I don't think overloading `operator==` for anything other than exact equality is allowed by https://google.github.io/styleguide/cppguide.html#Operator_Overloading.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by jwnimmer-tri** [comment]
*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg0-6u8CARPdLsQiaK1:-Kg1-Ou6NOKe2iqeBFaO:blol3gi) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L1)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

you mean plan_eval_utils.h? it seems fine to me, maybe i read it wrong? the other stuff seems pretty bad.. 
</blockquote></details>

In my report, it was showing the called methods around 100% covered (green), but around 50% of the methods were never called (and so had no markings at all).  Maybe I read the report wrong?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by siyuanfeng-tri** [comment]
*[drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h, line 1 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-Kg0-6u8CARPdLsQiaK1:-Kg10R1XRnE5_wBurxwv:bwxkaxq) ([raw file](https://github.com/robotlocomotion/drake/blob/92081dfd1197725ee3eec804f10b952207a872a6/drake/examples/QPInverseDynamicsForHumanoids/plan_eval/plan_eval_utils.h#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

In my report, it was showing the called methods around 100% covered (green), but around 50% of the methods were never called (and so had no markings at all).  Maybe I read the report wrong?
</blockquote></details>

ah i see what you are talking about. i added more tests, they should cover everything now. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585)*

**Comment by jwnimmer-tri** [comment]
+(status: curate commits before merging) 
:lgtm:

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5585#-:-Kg1WklrK4Vn8NEl9fr-:b-kmrex5)*
